File: lib/features/player/data/player_item_utils.dart
import 'package:clashkingapp/core/services/game_data_service.dart';
import 'package:clashkingapp/features/player/models/player_item.dart';

typedef PlayerItemFactory<T extends PlayerItem> =
    T Function({
      required String name,
      required int level,
      required int maxLevel,
      required bool isUnlocked,
      Map<String, dynamic>? meta,
      Map<String, dynamic>? rawJson,
    });

List<T> generateCompleteItemList<T extends PlayerItem>({
  required List<dynamic>? jsonList,
  required Map<String, dynamic> gameData,
  required PlayerItemFactory<T> factory,
  bool Function(String itemName, dynamic jsonItem)? nameMatcher,
}) {
  final Map<String, Map<String, dynamic>> allItems = Map.fromEntries(
    gameData.entries.map(
      (e) => MapEntry(e.key, e.value as Map<String, dynamic>),
    ),
  );

  return allItems.entries.map((entry) {
    final key = entry.key;
    final meta = entry.value;
    final name = meta['name']?.toString() ?? key;
    final owned = jsonList?.firstWhere(
      (x) =>
          nameMatcher?.call(key, x) ??
          nameMatcher?.call(name, x) ??
          x['name'] == name,
      orElse: () => null,
    );

    return factory(
      name: name,
      level: owned?['level'] ?? 0,
      maxLevel: owned?['maxLevel'] ?? meta['maxLevel'] ?? 0,
      isUnlocked: owned != null,
      meta: meta,
      rawJson: owned,
    );
  }).toList();
}

Map<String, dynamic> filterGameData(
  Map<String, dynamic>? data,
  bool Function(String key, Map<String, dynamic> value) predicate,
) {
  if (data == null) return {};

  final filteredEntries = <MapEntry<String, dynamic>>[];
  for (final entry in data.entries) {
    final value = entry.value;
    if (value is! Map) {
      continue;
    }

    final typedValue = Map<String, dynamic>.from(value);
    if (typedValue['is_seasonal'] == true) {
      continue;
    }

    if (predicate(entry.key, typedValue)) {
      filteredEntries.add(MapEntry(entry.key, typedValue));
    }
  }

  return Map.fromEntries(filteredEntries);
}

Map<String, dynamic> filterSpellGameData(Map<String, dynamic>? data) {
  return filterGameData(data, (_, _) => true);
}

/// Returns the highest item level unlockable at [thLevel] based on bundle metadata.
/// Returns 0 if the item has no levels or is not available at that TH level.
int maxLevelForTH(
  Map<String, dynamic>? meta,
  int thLevel, {
  int? maxTownHallLevel,
  int? itemMaxLevel,
}) {
  if (meta == null || thLevel <= 0) return 0;
  final levels = meta['levels'];
  if (levels is! List || levels.isEmpty) return 0;
  int maxLevel = 0;
  for (final entry in levels) {
    if (entry is Map) {
      final requiredTH = entry['required_townhall'];
      final level = entry['level'];
      if (requiredTH is num && level is num && requiredTH <= thLevel) {
        if (level > maxLevel) maxLevel = level.toInt();
      }
    }
  }

  final declaredMaxLevel =
      itemMaxLevel ?? (meta['maxLevel'] as num?)?.toInt() ?? 0;
  final effectiveMaxTownHallLevel =
      maxTownHallLevel ?? GameDataService.getMaxTownHallLevel();
  if (effectiveMaxTownHallLevel > 0 &&
      thLevel >= effectiveMaxTownHallLevel &&
      declaredMaxLevel > maxLevel) {
    return declaredMaxLevel;
  }

  return maxLevel;
}

int maxLevelForItemAtTH(PlayerItem item, int thLevel, {int? maxTownHallLevel}) {
  final thMax = maxLevelForTH(
    item.meta,
    thLevel,
    maxTownHallLevel: maxTownHallLevel,
    itemMaxLevel: item.maxLevel,
  );

  if (thMax > 0) return thMax;
  return item.maxLevel;
}

class UpgradeResourceAmount {
  final String key;
  final num amount;

  const UpgradeResourceAmount(this.key, this.amount);
}

class UpgradeRemainingSummary {
  final int targetLevel;
  final int levelsRemaining;
  final int seconds;
  final List<UpgradeResourceAmount> resources;

  const UpgradeRemainingSummary({
    required this.targetLevel,
    required this.levelsRemaining,
    required this.seconds,
    required this.resources,
  });

  bool get isComplete => levelsRemaining <= 0;
}

UpgradeRemainingSummary calculateRemainingUpgradeSummary(
  PlayerItem item, {
  required int targetLevel,
}) {
  final meta = item.meta;
  if (meta == null || targetLevel <= 0 || item.level >= targetLevel) {
    return UpgradeRemainingSummary(
      targetLevel: targetLevel,
      levelsRemaining: 0,
      seconds: 0,
      resources: const [],
    );
  }

  final costs = <String, num>{};
  var totalSeconds = 0;
  final firstUpgradeLevel = item.level <= 0 ? 1 : item.level;
  final levels = meta['levels'];
  final levelStats = <int, Map<String, dynamic>>{
    if (levels is List)
      for (final entry in levels.whereType<Map>())
        if (entry['level'] is num)
          (entry['level'] as num).toInt(): Map<String, dynamic>.from(entry),
  };

  for (var level = firstUpgradeLevel; level < targetLevel; level++) {
    final stats = levelStats[level];
    if (stats == null) continue;

    totalSeconds += (stats['upgrade_time'] as num?)?.toInt() ?? 0;
    final upgradeCost = stats['upgrade_cost'];

    if (upgradeCost is Map) {
      for (final entry in upgradeCost.entries) {
        final amount = entry.value is num ? entry.value as num : 0;
        if (amount <= 0) continue;
        final key = normalizeResource(entry.key.toString());
        costs[key] = (costs[key] ?? 0) + amount;
      }
    } else {
      final amount = upgradeCost is num ? upgradeCost : 0;
      if (amount <= 0) continue;
      final key = normalizeResource(
        meta['upgrade_resource']?.toString() ?? 'resource',
      );
      costs[key] = (costs[key] ?? 0) + amount;
    }
  }

  final resources =
      costs.entries
          .map((entry) => UpgradeResourceAmount(entry.key, entry.value))
          .where((resource) => resource.amount > 0)
          .toList()
        ..sort(
          (a, b) =>
              resourceSortWeight(a.key).compareTo(resourceSortWeight(b.key)),
        );

  return UpgradeRemainingSummary(
    targetLevel: targetLevel,
    levelsRemaining: (targetLevel - item.level).clamp(0, targetLevel),
    seconds: totalSeconds,
    resources: resources,
  );
}

Map<String, dynamic>? findLevelStats(Map<String, dynamic>? meta, int level) {
  if (meta == null) return null;
  final levels = meta['levels'];
  if (levels is! List) return null;
  for (final entry in levels) {
    if (entry is Map && entry['level'] == level) {
      return Map<String, dynamic>.from(entry);
    }
  }
  return null;
}

String normalizeResource(String value) {
  return value.trim().toLowerCase().replaceAll(' ', '_').replaceAll('-', '_');
}

int resourceSortWeight(String key) {
  if (key.contains('gold') && !key.contains('builder')) return 0;
  if (key.contains('elixir') && !key.contains('dark')) return 1;
  if (key.contains('dark')) return 2;
  if (key.contains('builder') && key.contains('gold')) return 3;
  if (key.contains('builder') && key.contains('elixir')) return 4;
  if (key.contains('shiny')) return 5;
  if (key.contains('glowy')) return 6;
  if (key.contains('starry')) return 7;
  return 99;
}

File: expo/src/features/player/data/player-item-utils.ts
import { getMaxTownHallLevel } from '@/core/game-data/game-data-service';
import { isRecord, type JsonRecord } from '../models/parsing';
import type { PlayerItem, RawPlayerItemInput } from '../models/player-items';

export function generateCompleteItemList<T>(input: {
  jsonList?: unknown[] | null;
  gameData: JsonRecord;
  factory: (value: RawPlayerItemInput) => T;
  nameMatcher?: (itemName: string, jsonItem: JsonRecord) => boolean;
}): T[] {
  return Object.entries(input.gameData)
    .filter((entry): entry is [string, JsonRecord] => isRecord(entry[1]))
    .map(([key, meta]) => {
      const name = meta.name == null ? key : String(meta.name);
      const owned = input.jsonList?.find(
        (item) =>
          isRecord(item) &&
          (input.nameMatcher?.(key, item) ?? input.nameMatcher?.(name, item) ?? item.name === name),
      );
      const raw = isRecord(owned) ? owned : null;
      return input.factory({
        name,
        level: typeof raw?.level === 'number' ? Math.trunc(raw.level) : 0,
        maxLevel:
          typeof raw?.maxLevel === 'number'
            ? Math.trunc(raw.maxLevel)
            : typeof meta.maxLevel === 'number'
              ? Math.trunc(meta.maxLevel)
              : 0,
        isUnlocked: raw !== null,
        meta,
        rawJson: raw,
        superTroopIsActive: raw?.superTroopIsActive === true,
      });
    });
}
export function filterGameData(
  data: unknown,
  predicate: (key: string, value: JsonRecord) => boolean,
): JsonRecord {
  if (!isRecord(data)) return {};
  return Object.fromEntries(
    Object.entries(data).filter(
      (entry): entry is [string, JsonRecord] =>
        isRecord(entry[1]) && entry[1].is_seasonal !== true && predicate(entry[0], entry[1]),
    ),
  );
}
export const filterSpellGameData = (data: unknown) => filterGameData(data, () => true);
export function maxLevelForTH(
  meta: JsonRecord | null | undefined,
  thLevel: number,
  options: { maxTownHallLevel?: number; itemMaxLevel?: number } = {},
): number {
  if (!meta || thLevel <= 0 || !Array.isArray(meta.levels) || meta.levels.length === 0) return 0;
  let max = 0;
  for (const entry of meta.levels)
    if (
      isRecord(entry) &&
      typeof entry.required_townhall === 'number' &&
      typeof entry.level === 'number' &&
      entry.required_townhall <= thLevel
    )
      max = Math.max(max, Math.trunc(entry.level));
  const declared =
      options.itemMaxLevel ?? (typeof meta.maxLevel === 'number' ? Math.trunc(meta.maxLevel) : 0),
    maxTh = options.maxTownHallLevel ?? getMaxTownHallLevel();
  return maxTh > 0 && thLevel >= maxTh && declared > max ? declared : max;
}
export function maxLevelForItemAtTH(item: PlayerItem, thLevel: number, maxTownHallLevel?: number) {
  return (
    maxLevelForTH(item.meta, thLevel, { maxTownHallLevel, itemMaxLevel: item.maxLevel }) ||
    item.maxLevel
  );
}
export interface UpgradeResourceAmount {
  key: string;
  amount: number;
}
export interface UpgradeRemainingSummary {
  targetLevel: number;
  levelsRemaining: number;
  seconds: number;
  resources: readonly UpgradeResourceAmount[];
  isComplete: boolean;
}
export function calculateRemainingUpgradeSummary(
  item: PlayerItem,
  targetLevel: number,
): UpgradeRemainingSummary {
  if (!item.meta || targetLevel <= 0 || item.level >= targetLevel)
    return { targetLevel, levelsRemaining: 0, seconds: 0, resources: [], isComplete: true };
  const stats = new Map<number, JsonRecord>();
  if (Array.isArray(item.meta.levels))
    for (const entry of item.meta.levels)
      if (isRecord(entry) && typeof entry.level === 'number')
        stats.set(Math.trunc(entry.level), entry);
  let seconds = 0;
  const costs: Record<string, number> = {};
  for (let level = item.level <= 0 ? 1 : item.level; level < targetLevel; level++) {
    const row = stats.get(level);
    if (!row) continue;
    seconds += typeof row.upgrade_time === 'number' ? Math.trunc(row.upgrade_time) : 0;
    if (isRecord(row.upgrade_cost)) {
      for (const [key, value] of Object.entries(row.upgrade_cost)) {
        const amount = typeof value === 'number' ? value : 0;
        if (amount > 0) {
          const normalized = normalizeResource(key);
          costs[normalized] = (costs[normalized] ?? 0) + amount;
        }
      }
    } else {
      const amount = typeof row.upgrade_cost === 'number' ? row.upgrade_cost : 0;
      if (amount > 0) {
        const key = normalizeResource(String(item.meta.upgrade_resource ?? 'resource'));
        costs[key] = (costs[key] ?? 0) + amount;
      }
    }
  }
  const resources = Object.entries(costs)
    .map(([key, amount]) => ({ key, amount }))
    .sort((a, b) => resourceSortWeight(a.key) - resourceSortWeight(b.key));
  const levelsRemaining = Math.max(0, Math.min(targetLevel, targetLevel - item.level));
  return { targetLevel, levelsRemaining, seconds, resources, isComplete: levelsRemaining <= 0 };
}
export function findLevelStats(
  meta: JsonRecord | null | undefined,
  level: number,
): JsonRecord | null {
  if (!Array.isArray(meta?.levels)) return null;
  return (
    (meta.levels.find((entry) => isRecord(entry) && entry.level === level) as
      JsonRecord | undefined) ?? null
  );
}
export const normalizeResource = (value: string) =>
  value.trim().toLowerCase().replaceAll(' ', '_').replaceAll('-', '_');
export function resourceSortWeight(key: string) {
  if (key.includes('gold') && !key.includes('builder')) return 0;
  if (key.includes('elixir') && !key.includes('dark')) return 1;
  if (key.includes('dark')) return 2;
  if (key.includes('builder') && key.includes('gold')) return 3;
  if (key.includes('builder') && key.includes('elixir')) return 4;
  if (key.includes('shiny')) return 5;
  if (key.includes('glowy')) return 6;
  if (key.includes('starry')) return 7;
  return 99;
}
